### PR TITLE
.get_undecorated_callback() bugfix

### PR DESCRIPTION
**Pull request by Bystroushaak**
I've found obscure error triggered in case of decorators with multiple parameters.

Now, the ``.get_undecorated_callback()`` method just picks first attribute, which doesn't have to be function, but also one of the parameters of decorator with multiple parameters.

For example, this decorator will trigger the bug:

```python
def database_decorator(fn=None, filename="default.sqlite", table="default"):
    """
    Open connection to the database and offer it to the wrapped function.

    So far, there is no database, only primitive JSON load/store system.
    """
    def database_decorator_wrapper(fn):
        @wraps(fn)
        def database_wrapper(*args, **kwargs):
            path = filename
            if not os.path.isabs(path):
                path = os.path.join(settings.DATABASE_PATH, path)

            # load database
            db = SqliteDict(
                path,
                # autocommit=True
            )

            # load table
            table_ref = db.get(table, {})

            # put table to function parameters
            kwargs["db"] = table_ref

            out = fn(*args, **kwargs)

            # save database
            db[table] = table_ref
            db.commit()

            return out

        return database_wrapper

    if fn:  # python decorator with optional parameters bukkake
        return database_decorator_wrapper(fn)

    return database_decorator_wrapper
```

My patch just looks for cases when the ``func`` is not ``FunctionType`` instance (which it should be) and then goes thru all the parameters and pick first ``FunctionType`` instance.

### CODE REVIEW

**Comment by avelino** [comment]
Ideally, you write a test to cover implementation (refactoring) proposal.

**Comment by Bystroushaak** [comment]
Well, it didn't break anything, but I will add the tests.

**Comment by Bystroushaak** [comment]
This bug is hard (at least for me) to trigger during test, because it is triggered at runtime.

This is the error message I get for my code:

```python
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/bottle.py", line 862, in _handle
    return route.call(**args)
  File "/usr/local/lib/python2.7/dist-packages/bottle.py", line 1732, in wrapper
    rv = callback(*a, **ka)
  File "/usr/local/lib/python2.7/dist-packages/bottle_gui/bottle_gui.py", line 372, in root
    filter(lambda x: x.path not in BLACKLIST, list_routes())
  File "/usr/local/lib/python2.7/dist-packages/bottle_gui/bottle_gui.py", line 250, in list_routes
    bottle.default_app().routes
  File "/usr/local/lib/python2.7/dist-packages/bottle_gui/bottle_gui.py", line 243, in <lambda>
    args=r.get_callback_args(),
  File "/usr/local/lib/python2.7/dist-packages/bottle.py", line 551, in get_callback_args
    return getargspec(self.get_undecorated_callback())[0]
  File "/usr/lib/python2.7/inspect.py", line 816, in getargspec
    raise TypeError('{!r} is not a Python function'.format(func))
TypeError: './ticker_db.sqlite' is not a Python function
```

when the function is defined as:

```python
DB_FN = "./key_val_db.sqlite"
DB_TABLE = "ticker"

...

@post(SERVICE_URL)
@database_decorator(filename=DB_FN, table=DB_TABLE)
@json_to_params
def save(db, key, val):
    """
    To `key` save `val`.
    """
    db[key] = val
```

Where ``database_decorator()`` was listed in first comment and ``json_to_params()`` is defined here: https://github.com/Bystroushaak/bottle-rest/blob/master/src/bottle_rest/bottle_rest.py#L109

Then when I try to access ``.get_callback_args()`` at **runtime**, the error is triggered. I patched the code with the patch from pull request and the error is fixed.

I am currently trying to write failing test for unpatched version of bottle, but no luck so far.

**Comment by Bystroushaak** [comment]
Ah, finally.

**Comment by Bystroushaak** [comment]
Here is the test body (from ``test_route.py``):

```python
    def test_callback_inspection_multiple_args(self):
        # decorator with multiple arguments
        def d2(fn=None, filename="default.sqlite", table="default"):
            def d(fn):
                def w(*args, **kwargs):
                    kwargs["a"] = filename
                    kwargs["b"] = table
                    return fn(*args, **kwargs)
                return w

            # support for calling this decorator without arguments
            if fn:
                return d(fn)

            return d

        def pretty_dump(fn):
            def pretty_dump_wrapper(*args, **kwargs):
                return fn(*args, **kwargs)

            return pretty_dump_wrapper

        # multiple decorators
        @bottle.get("somepath")
        @d2(filename='foo', table='bar')
        @pretty_dump
        def x(a, b):
            return

        route = bottle.Route(None, None, None, x)
        self.assertEqual(set(route.get_callback_args()), set(['a', 'b']))
```

This fails at ``master`` branch, but not at patched version.

**Comment by foxbunny** [comment]
You should have used a simplified decorator for the test. This one makes the test way too hard to follow.

**Comment by Bystroushaak** [comment]
Simplified decorator doesn't trigger the bug imho. At least I couldnt trigger the bug when I tried.

**Comment by foxbunny** [comment]
No, I mean how is `filename="default.sqlite", table="default"` relevant? Could've been just `foo, bar`. Etc.

**Comment by Bystroushaak** [comment]
Oh, yes. I just cut and pasted parts of my code, where I yestearday found the bug.

**Comment by Bystroushaak** [comment]
I will minimize it and also send the bugfix for python 3, which is now failing in travis.

**Comment by foxbunny** [comment]
You should strip the decorator down to just the parts that exhibit the desired behavior without actually handling args/kwargs, etc, and then test with only that code, so you can rule out any foul in your handling of the args.

**Comment by Bystroushaak** [comment]
The bug shows only when I use kwargs.

**Comment by foxbunny** [comment]
For example, does the bug manifest itself if you **don't** do this:

```
kwargs["a"] = filename
kwargs["b"] = table
```

And does it manifest itself regardless of what the kwargs are named? And so on. Too much noise that doesn't seem too relevant.

**Comment by Bystroushaak** [comment]
Thanks for your feedback. I will try to change the code.

**Comment by Bystroushaak** [comment]
I think, that it will be better, if I change the code and then send new pull request, this one is too long and messy. Sorry.
